core/src/kmx/kmx_plus.cpp

```diff
@@ -92,105 +92,353 @@ validate_section_name(KMX_DWORD ident) {
-    return nullptr;
+    return false;
   }
-  if (length < LDML_LENGTH_HEADER) {
-    DebugLog("length < LDML_LENGTH_HEADER");
-    assert(false);
-    return nullptr;
+  if(fileVersion == LDML_KMXPLUS_VERSION_17) {
+    const COMP_KMXPLUS_HEADER_17 *all = reinterpret_cast<const COMP_KMXPLUS_HEADER_17 *>(data);
+    if (!all->valid(length)) {
+      DebugLog("header failed validation");
+      assert(false);
+      return false;
+    }
+
+    if (all->ident != ident) {
+      DebugLog("header had wrong section id");
+      assert(false);
+      return false;
+    }
+
+    out.set(LDML_KMXPLUS_VERSION_17, all->ident, all->size);
+  } else {
+    const COMP_KMXPLUS_HEADER_19 *all = reinterpret_cast<const COMP_KMXPLUS_HEADER_19 *>(data);
+    if (!all->valid(length)) {
+      DebugLog("header failed validation");
+      assert(false);
+      return false;
+    }
+
+    if(ident == LDML_SECTIONID_SECT) {
+      if(all->ident != LDML_SECTIONID_SEC2) {
+        DebugLog("header had SECT but expected SEC2");
+        assert(false);
+        return false;
+      }
+    } else if (all->ident != ident) {
+      DebugLog("header had wrong section id");
+      assert(false);
+      return false;
+    }
+
+    out.set(LDML_KMXPLUS_VERSION_19, all->ident, all->size, all->version);
   }
-  const COMP_KMXPLUS_HEADER *all = reinterpret_cast<const COMP_KMXPLUS_HEADER *>(data);
-  if (!all->valid(length)) {
-    DebugLog("header failed validation");
+
+  return true;
+}
+
+/**
+ * @brief Determines the overall version of the file based on the first four bytes:
+ *          'sect' = v17 file
+ *          'sec2' = v19+ file
+ *          other = invalid file
+ * A v19+ file will include an additional version field for each section.
+ * @param data
+ * @param length
+ * @return KMX_DWORD  a LDML_KMXPLUS_VERSION_ value, or 0 on error
+ */
+KMX_DWORD determine_file_version_from_bytes(const uint8_t* data, KMX_DWORD length) {
+  if(length < sizeof(COMP_KMXPLUS_HEADER_17)) {
+    DebugLog("data is too short");
     assert(false);
-    return nullptr;
+    return 0;
   }
-  if (all->ident != ident) {
-    DebugLog("header had wrong section id");
-    assert(false);
-    return nullptr;
+  COMP_KMXPLUS_HEADER_17 const* header = reinterpret_cast<COMP_KMXPLUS_HEADER_17 const *>(data);
+  if(header->ident == COMP_KMXPLUS_SECT::IDENT) {
+    return LDML_KMXPLUS_VERSION_17;
+  }
+  if(header->ident == COMP_KMXPLUS_SECT::IDENT_V19) {
+    return LDML_KMXPLUS_VERSION_19;
   }
-  return all;
+  DebugLog("initial section is unrecognized");
+  assert(false);
+  return 0;
 }
-
 /**
  * @brief Accessor for a section based on bytes
  *
  * @tparam T
  * @param data input pointer to the section to be validated. If nullptr, 'sect' will be nullptr, but will return true (valid).
- * @param length length of input data. This function will not read past the end of this length.
+ * @param fileLength length of input data. This function will not read past the end of this length.
```

Comment:
Why the name change it isn't the file lenght is it? it is the length of the data, or section length which is not the whole file?

Reply:
No, this reads a whole section from the raw binary data, starting from the first byte of the KMX+ data (`sect` or `sec2`). So `fileLength` means the entire length of the KMX+ data (which may be embedded in a KMX file, but does not have to be, by design).